spark/src/main/com/intel/genomicsdb/GenomicsDBInputFormat.scala

```diff
@@ -0,0 +1,193 @@
+    * and creates one-to-one corresponding input splits. The partitions are in Spark
+    * are created based on this number [[com.intel.genomicsdb.GenomicsDBRDD]]
+    * Currently, there is 1:1 relation to GenomicsDB data partition per node and
+    * input split. This might change in future!!
+    *
+    * @return  List of input splits
+    */
+  @DeveloperApi
+  @Override
+  def getSplits(runningAsService: Boolean = false): List[InputSplit] = {
+
+    // When GenomicsDB runs as a service, get the partition
+    // information from the server, else, read from the hosts
+    // file provided in the configuration object
+    if (runningAsService) {
+      // @todo: do nothing yet
+      createConnection()
+      null
+    } else {
+      val hostList = gConf.getHosts
+      val splits = new Array[InputSplit](hostList.size)
+
+      for (i <- 0 until hostList.size) {
+        splits(i) = GenomicsDBInputSplit(Array(hostList.get(i)), gConf)
+      }
+      splits.toList
+    }
+  }
+
+  def createRecordReader(
+    inputSplit: InputSplit,
+    taskContext: TaskContext,
+    flag: Int): RecordReader[LongWritable, VCONTEXT] = {
+
+    //val config = taskAttemptContext.getConfiguration
+
+    if (!loaderJsonFile.toString.isEmpty && !queryJsonFile.toString.isEmpty) {
+      featureReader = new GenomicsDBFeatureReader(loaderJsonFile.toString,
+        queryJsonFile.toString, new BCF2Codec()).asInstanceOf
+    } else {
+      val tiledbWorkspace = gConf.getTileDBWorkspace
+      val arrayName = gConf.getTileDBArrayName
+      featureReader = new GenomicsDBFeatureReader(tiledbWorkspace.toString, arrayName,
+        gConf.getReferenceGenome, gConf.getVCFHeaderFile, new BCF2Codec()).asInstanceOf
+    }
+    header = featureReader.getHeader
+    recordReader = new GenomicsDBRecordReader[VCONTEXT, SOURCE](inputSplit.asInstanceOf, gConf)
+    recordReader
+  }
+
+  def getSplits(jobContext: JobContext): JavaList[InputSplit] = {
+    // do nothing. never used
+    List.empty.asInstanceOf
+  }
+
+  /**
+    * Return the header from the gVCF from any partition.
+    * Header is same across all gVCF files
+    *
+    * @return
+    */
+  def getHeader: Object = header
+
+
+  /**
+    * GenomicsDB will be setup as a service to accommodate thousands of compute nodes
+    * retreive data from a smaller set of data nodes. In such a scenario, need to create
+    * a connection prior to retrieving records and do not assume data will be local
+    */
+  def createConnection() = {
+    try {
+      val transport : TTransport = new TSocket(gConf.getHosts.get(0), gConf.getGenomicsDBPort)
+      transport.open()
+
+      val protocol: TProtocol = new TBinaryProtocol(transport)
+      //val client: GenomicsDBClient = new GenomicsDBService.Client(protocol)
+      //perform(client);
+      transport.close();
+    } catch {
+      case e: TException =>
+        e.printStackTrace()
+    }
+  }
+
+  def getRecordReader: GenomicsDBRecordReader[VCONTEXT, SOURCE] = recordReader
```

Comment:
The interface does not mandate a parameter, so I use the split passed to the createRecordReader.